File: backend/internal/domain/customs.go
package domain

import "time"

type CustomsDocumentCheck struct {
	ID                 string                     `json:"id"`
	ShipmentID         string                     `json:"shipment_id"`
	Source             string                     `json:"source"`
	DocType            string                     `json:"doc_type"`
	OriginalFileName   string                     `json:"original_file_name"`
	ContentType        string                     `json:"content_type"`
	SizeBytes          int64                      `json:"size_bytes"`
	SHA256Hash         string                     `json:"sha256_hash"`
	VerificationStatus DocumentVerificationStatus `json:"verification_status,omitempty"`
	HashMatches        bool                       `json:"hash_matches"`
	HashStatus         string                     `json:"hash_status"`
	UploadedAt         time.Time                  `json:"uploaded_at"`
}

type CustomsShipmentDetail struct {
	Shipment     Shipment               `json:"shipment"`
	Documents    []CustomsDocumentCheck `json:"documents"`
	Events       []ShipmentEvent        `json:"events"`
	ReleaseReady bool                   `json:"release_ready"`
}

File: frontend/src/lib/customs.ts
import { API_BASE, apiFetch } from "@/lib/api";
import type {
  DocumentVerificationStatus,
  Shipment,
  ShipmentEvent,
} from "@/lib/shipments";

export type CustomsDocumentCheck = {
  id: string;
  shipment_id: string;
  source: "IMPORTER" | "SELLER";
  doc_type: string;
  original_file_name: string;
  content_type: string;
  size_bytes: number;
  sha256_hash: string;
  verification_status?: DocumentVerificationStatus;
  hash_matches: boolean;
  hash_status: string;
  uploaded_at: string;
};

export type CustomsShipmentDetail = {
  shipment: Shipment;
  documents: CustomsDocumentCheck[];
  events: ShipmentEvent[];
  release_ready: boolean;
};

export async function listCustomsArrivedShipments(
  token: string,
): Promise<Shipment[]> {
  const res = await apiFetch<{ items?: Shipment[] | null }>(
    "/api/v1/customs/shipments/arrived",
    { token },
  );
  return Array.isArray(res.items) ? res.items : [];
}

export async function getCustomsShipment(
  token: string,
  shipmentID: string,
): Promise<CustomsShipmentDetail> {
  const detail = await apiFetch<CustomsShipmentDetail>(
    `/api/v1/customs/shipments/${shipmentID}`,
    { token },
  );
  return {
    ...detail,
    documents: Array.isArray(detail.documents) ? detail.documents : [],
    events: Array.isArray(detail.events) ? detail.events : [],
  };
}

export async function grantCustomsRelease(
  token: string,
  shipmentID: string,
): Promise<CustomsShipmentDetail> {
  const detail = await apiFetch<CustomsShipmentDetail>(
    `/api/v1/customs/shipments/${shipmentID}/release`,
    {
      method: "POST",
      token,
    },
  );
  return {
    ...detail,
    documents: Array.isArray(detail.documents) ? detail.documents : [],
    events: Array.isArray(detail.events) ? detail.events : [],
  };
}

export async function fetchClearanceCertificate(
  token: string,
  shipmentID: string,
): Promise<Blob> {
  const res = await fetch(
    `${API_BASE}/api/v1/customs/shipments/${shipmentID}/certificate`,
    {
      headers: { Authorization: `Bearer ${token}` },
    },
  );
  if (!res.ok) {
    const text = await res.text().catch(() => "");
    throw new Error(text || `Request failed (${res.status})`);
  }
  return res.blob();
}
